### PR TITLE
feat: Add strip_optional method to TypeView.

### PR DESCRIPTION
**Pull request by DanCardin**
## Description
Adds a method which allows removing the "optional" component of an input type that `is_optional`

## Closes
Fixes https://github.com/litestar-org/type-lens/issues/8

### CODE REVIEW

File: type_lens/type_view.py
Comment (peterschutt) [inline type_lens/type_view.py:141]: What about `as_non_optional` or `ensure_non_optional`?

I suppose there's two things for the name to convey, if possible. 1. that it can be called on any type, not just optional unions, and 2. that it returns a different instance.

Comment (DanCardin) [reply type_lens/type_view.py:141]: `strip_optional`?

Comment (peterschutt) [reply type_lens/type_view.py]: I'm good with `strip_optional` :+1:

File: type_lens/type_view.py
Comment (peterschutt) [inline type_lens/type_view.py:157]:
```suggestion
        args = tuple(a for a in self.args if a is not NoneType)
```
No need to create the list first.

**Review by peterschutt — APPROVED** [review]

File: type_lens/type_view.py
Comment (provinzkraut) [inline type_lens/type_view.py:158]: I know I'm a bit late to the party but I think `without_none_union` would make more sense in overall. For one, it fits the wording of the other properties better (`is_`, `has_`, etc.), and more importantly, it's closer to what's happening under the hood (same is true for `is_optional`, which probably should be something like `is_none_union`).

@peterschutt Wdyt about making these changes?

Comment (peterschutt) [reply type_lens/type_view.py:158]: I don't mind the fact that we use the term "optional" to mean "union with `None`", because it has that meaning assigned to it from `typing.Optional`. What about `without_optional`?

Comment (provinzkraut) [reply type_lens/type_view.py:158]: > because it has that meaning assigned to it from `typing.Optional`. What about `without_optional`?

That *is* going away though and is being replaced by `<type> | None`, that's what I was thinking. Might be a bit of a long-term thought here but Python 3.9 *is* going to be EOL by the end of next year, so by then it's probably safe to assume `Optional` usage isn't the default anymore really and not part of the supported versions anyway.

> What about `without_optional`?

Sounds good!

Comment (DanCardin) [reply type_lens/type_view.py:158]: the optional verbiage isn't deprecated or going away, even if there's now syntax to avoid it.

The thing that concerns me about "without_" is that it maybe implies more than happens. `Optional[list[Optional[int]]]` would yield `list[Optional[int]]` so it's not like it's..."without" it anymore. it's doing a one-level strip of the optional that may or may not exist.

> For one, it fits the wording of the other properties better (is_, has_, etc.)
And I feel like that's not particularly meaningful personally. Every other method is a boolean check where a passive verb(?) fits more, imo.

I'm pro active `strip_` verb personally. Or perhaps something like `remove_` (which i suppose lines up with `"".removeprefix`).

Comment (provinzkraut) [reply type_lens/type_view.py:158]: > the optional verbiage isn't deprecated or going away, even if there's now syntax to avoid it.

True, but it being less common plus it not being represented as such (i.e. `Optional` is just `| None` internally anyway), would mean that there's less of a strong connection between what an "optional" type is.

> Or perhaps something like `remove_` (which i suppose lines up with `"".removeprefix`).

I feel like `remove_` as well would imply that it removes inner `Optional`s as well.
In this case I think `strip_` is the only thing that makes clear that it's only removing the outer layer. 

However, I think `without_` communicates another aspect better: That it returns a new thing *without* this specific property. `strip_` on the other hand could also indicate that the instance is being mutated. This holds true for the stdlib as well, e.g. `list.sort()` and `dict.clear()` mutate in place. It's only on immutables that this verbiage is used to indicate the creation of a new instance, e.g. `str.strip`.

Comment (peterschutt) [reply type_lens/type_view.py:158]: > It's only on immutables that this verbiage is used to indicate the creation of a new instance, e.g. str.strip.

These are intended to be immutable - at least that's what I was going for by marking everything as `Final`.

Comment (peterschutt) [reply type_lens/type_view.py:158]: I think the overloading of the term Optional to mean union with none is unfortunate, and would be happy to see it go away. So how about `strip_none_union`?

Comment (provinzkraut) [reply type_lens/type_view.py:158]: > I think the overloading of the term Optional to mean union with none is unfortunate, and would be happy to see it go away. So how about `strip_none_union`? And we'd rename `is_optional` to `is_none_union`.

Sounds good to me!

Comment (DanCardin) [reply type_lens/type_view.py:158]: i mean, i'm not the library owner, but i still think going against the verbiage of python itself, [typing_inspect](https://github.com/ilevkivskyi/typing_inspect) (maintained by a mypy maintainer seemingly), and mypy/pyright at a glance is unnecessary.

I feel like "optional" is very much a more obvious sounding name than "none_union" given the history and everything else

Comment (guacs) [reply type_lens/type_view.py:158]: I agree with @DanCardin on this. I feel like `strip_optional` makes sense over `strip_none_union.` Also, like he said while the `Optional` type might be going away, I think it's pretty common for people to mention `str | None` types as an optional string etc.

Comment (provinzkraut) [reply type_lens/type_view.py:158]: > i still think going against the verbiage of python itself, [typing_inspect](https://github.com/ilevkivskyi/typing_inspect) (maintained by a mypy maintainer seemingly), and mypy/pyright at a glance is unnecessary.

Maybe. Sometimes it's best to go with the flow, but sometimes it's also worth going in the opposite direction. Personally, I think `Optional` to mean "a union with `None`" was a poor choice, as it creates much confusion. It might be worth considering to lead by example here how things could be improved. 

But then again, that's just my opinion. I'm not gonna block on this in any case :)

Comment (peterschutt) [reply type_lens/type_view.py:158]: I appreciate everyone weighing in on this.

I think that the choice to use the term "optional" to represent a union with `None` is unfortunate. My issue is that it has a meaning in lots of contexts, so when talking about typing through the prism of other domains, e.g., API endpoints/openapi schemas/function parameters, one has to be sure to convey what you mean by the term when using it in that context. E.g., making a parameter's type a union with `None`, doesn't make the parameter optional, the presence of a default value does. In an openapi context, optional could easily be conflated with a "not required" parameter. The old behavior of pydantic, where an explicit default for a `None` union wasn't required also blurred the lines around what "optional" means in both a typing, and parameter sense.

Given that we are split down the middle on this, and as I mentioned earlier in this thread - I recognize that it is a commonly accepted term in the python typing world, and as this is a library built around python's typing system, I'm OK with using it here.

So, I think status quo wins out, lets leave it as it is.

**Pull request merged**
